hilti/include/rt/types/set.h

```diff
@@ -25,33 +25,82 @@ namespace hilti::rt {
 namespace set {
 
 template<typename T>
-class SafeIterator
-    : public hilti::rt::detail::iterator::SafeIterator<Set<T>, typename Set<T>::SafeIterator, SafeIterator<T>> {
-public:
-    using Base = hilti::rt::detail::iterator::SafeIterator<Set<T>, typename Set<T>::SafeIterator, SafeIterator<T>>;
-    using Base::Base;
-};
+class Iterator {
+    using S = Set<T>;
+
+    std::weak_ptr<S*> _control;
+    typename S::V::iterator _iterator;
 
-template<typename T>
-class SafeConstIterator : public hilti::rt::detail::iterator::SafeIterator<const Set<T>, typename Set<T>::ConstIterator,
-                                                                           SafeConstIterator<T>> {
 public:
-    using Base =
-        hilti::rt::detail::iterator::SafeIterator<const Set<T>, typename Set<T>::ConstIterator, SafeConstIterator<T>>;
-    using Base::Base;
+    Iterator() = default;
+    Iterator(typename S::V::iterator iterator, const typename S::C& control)
```

Comment:
Seems like an internal API, could we make this protected or private, with `S` being a friend?